lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -1038,6 +1048,65 @@ data ApiAccountKey = ApiAccountKey
     } deriving (Eq, Generic, Show)
       deriving anyclass NFData
 
+data ApiSharedWalletPostDataFromMnemonics = ApiSharedWalletPostDataFromMnemonics
+    { name :: !(ApiT WalletName)
+    , mnemonicSentence :: !(ApiMnemonicT (AllowedMnemonics 'Shelley))
+    , mnemonicSecondFactor :: !(Maybe (ApiMnemonicT (AllowedMnemonics 'SndFactor)))
+    , passphrase :: !(ApiT (Passphrase "raw"))
+    , accountIndex :: !(ApiT DerivationIndex)
+    , paymentScriptTemplate :: !ScriptTemplate
+    , delegationScriptTemplate :: !(Maybe ScriptTemplate)
+    } deriving (Eq, Generic, Show)
+
+data ApiSharedWalletPostDataFromAccountPubX = ApiSharedWalletPostDataFromAccountPubX
+    { name :: !(ApiT WalletName)
+    , accountPublicKey :: !ApiAccountPublicKey
+    , accountIndex :: !(ApiT DerivationIndex)
+    , paymentScriptTemplate :: !ScriptTemplate
+    , delegationScriptTemplate :: !(Maybe ScriptTemplate)
+    } deriving (Eq, Generic, Show)
+
+newtype ApiSharedWalletPostData = ApiSharedWalletPostData
+    { wallet :: Either ApiSharedWalletPostDataFromMnemonics ApiSharedWalletPostDataFromAccountPubX
+    } deriving (Eq, Generic, Show)
+
+data ApiActiveSharedWallet = ApiActiveSharedWallet
+    { id :: !(ApiT WalletId)
+    , name :: !(ApiT WalletName)
+    , accountIndex :: !(ApiT DerivationIndex)
+    , addressPoolGap :: !(ApiT AddressPoolGap)
+    , passphrase :: !(Maybe ApiWalletPassphraseInfo)
+    , paymentScriptTemplate :: !ScriptTemplate
+    , delegationScriptTemplate :: !(Maybe ScriptTemplate)
+    , delegation :: !ApiWalletDelegation
+    , balance :: !ApiWalletBalance
+    , assets :: !ApiWalletAssetsBalance
+    , state :: !(ApiT SyncProgress)
+    , tip :: !ApiBlockReference
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
+
+data ApiPendingSharedWallet = ApiPendingSharedWallet
+    { id :: !(ApiT WalletId)
+    , name :: !(ApiT WalletName)
+    , accountIndex :: !(ApiT DerivationIndex)
+    , addressPoolGap :: !(ApiT AddressPoolGap)
+    , paymentScriptTemplate :: !ScriptTemplate
+    , delegationScriptTemplate :: !(Maybe ScriptTemplate)
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
+
+newtype ApiSharedWallet = ApiSharedWallet
+    { wallet :: Either ApiPendingSharedWallet ApiActiveSharedWallet
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
```

Comment:
Here also, it can be de-duplicated:

```suggestion
data ApiSharedWalletParams = ApiSharedWalletParams
    { accountIndex :: !(ApiT DerivationIndex)
    , paymentScriptTemplate :: !ScriptTemplate
    , delegationScriptTemplate :: !(Maybe ScriptTemplate)
    } deriving (Eq, Generic, Show)
      deriving anyclass NFData

data ApiPendingSharedWallet = ApiPendingSharedWallet
    { id :: !(ApiT WalletId)
    , name :: !(ApiT WalletName)
    , addressPoolGap :: !(ApiT AddressPoolGap)
    , sharedWalletParams :: !ApiSharedWalletParams
    } deriving (Eq, Generic, Show)
      deriving anyclass NFData

data ApiSharedWallet = ApiSharedWallet
    { baseWallet :: !ApiWallet
    , sharedWalletParams :: !ApiSharedWalletParams
    , active :: !Bool
    } deriving (Eq, Generic, Show)
      deriving anyclass NFData
```

Reply:
we can go even further here : 
```
2363 data ApiSharedWalletParams = ApiSharedWalletParams                                                                                                                                                            
2364     { id :: !(ApiT WalletId)                                                                                                                                                                                  
2365     , name :: !(ApiT WalletName)                                                                                                                                                                              
2366     , accountIndex :: !(ApiT DerivationIndex)                                                                                                                                                                 
2367     , paymentScriptTemplate :: !ScriptTemplate                                                                                                                                                                
2368     , delegationScriptTemplate :: !(Maybe ScriptTemplate)                                                                                                                                                     
2369     , addressPoolGap :: !(ApiT AddressPoolGap)                                                                                                                                                                
2370     } deriving (Eq, Generic, Show)                                                                                                                                                                            
2371       deriving anyclass NFData
```
ToJSON/FromJSON for ApiSharedWalletParams and ApiPendingSharedWallet will be trivial: 
```
2335 instance FromJSON ApiSharedWalletParams where                                                                                                                                                                 
2336     parseJSON = genericParseJSON defaultRecordTypeOptions                                                                                                                                                     
2337 instance ToJSON ApiSharedWalletParams where                                                                                                                                                                   
2338     toJSON = genericToJSON defaultRecordTypeOptions                                                                                                                                                           
2339                                                                                                                                                                                                               
2340 instance FromJSON ApiPendingSharedWallet where                                                                                                                                                                
2341     parseJSON = ApiPendingSharedWallet <$> (parseJSON @ApiSharedWalletParams)                                                                                                                                 
2342 instance ToJSON ApiPendingSharedWallet where                                                                                                                                                                  
2343     toJSON (ApiPendingSharedWallet content) =                                                                                                                                                                 
2344         toJSON content
```
but ApiActiveSharedWallet not so easy. And we will have difference between Api.Types and what is in swagger. Do we want this? (I feel it will be somewhere this discrepency between swagger and cardano-wallet core). We have it at shared state level already. Maybe it is good idea to leave it as it is? What you think?